pomerium/gateway/translate.go

```diff
@@ -0,0 +1,80 @@
+// Package gateway contains logic for converting Gateway API configuration into Pomerium
+// configuration.
+package gateway
+
+import (
+	"net/url"
+
+	"google.golang.org/protobuf/proto"
+
+	"github.com/pomerium/ingress-controller/model"
+	"github.com/pomerium/pomerium/config"
+	pb "github.com/pomerium/pomerium/pkg/grpc/config"
+)
+
+// TranslateRoutes converts from Gateway-defined routes to Pomerium route configuration protos.
+func TranslateRoutes(gc *model.GatewayHTTPRouteConfig) []*pb.Route {
+	// A single HTTPRoute may need to be represented using many Pomerium routes:
+	//  - An HTTPRoute may have multiple hostnames.
+	//  - An HTTPRoute may have multiple HTTPRouteRules.
+	//  - An HTTPRouteRule may have multiple HTTPRouteMatches.
+	// First we'll expand all HTTPRouteRules into "template" Pomerium routes, and then we'll
+	// repeat each "template" route once per hostname.
+	trs := templateRoutes(gc)
+
+	prs := make([]*pb.Route, 0, len(gc.Hostnames)*len(trs))
+	for _, h := range gc.Hostnames {
+		from := (&url.URL{
+			Scheme: "https",
+			Host:   string(h),
+		}).String()
+		for _, tr := range trs {
+			r := proto.Clone(tr).(*pb.Route)
+			r.From = from
+
+			// Skip any routes that fail to validate.
+			coreRoute, err := config.NewPolicyFromProto(r)
+			if err != nil || coreRoute.Validate() != nil {
+				continue
+			}
+
+			prs = append(prs, r)
+		}
+	}
+
+	return prs
+}
+
+// templateRoutes converts an HTTPRoute into zero or more Pomerium routes, ignoring hostname.
+func templateRoutes(gc *model.GatewayHTTPRouteConfig) []*pb.Route {
+	var prs []*pb.Route
+
+	rules := gc.Spec.Rules
+	for i := range rules {
+		rule := &rules[i]
+		pr := &pb.Route{}
+
+		// From the spec (near https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1.HTTPRoute):
+		// "Implementations MUST ignore any port value specified in the HTTP Host header while
+		// performing a match and (absent of any applicable header modification configuration) MUST
+		// forward this header unmodified to the backend."
+		pr.PreserveHostHeader = true
+
+		applyFilters(pr, rule.Filters)
+		applyBackendRefs(pr, gc, rule.BackendRefs)
```

Comment:
the following route would result in the following config, as we're making a route per rule

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: verify
  namespace: default
spec:
  parentRefs:
  - name: gateway-https
  hostnames:
  - "verify.localhost.pomerium.io"
  rules:
  - backendRefs:
    - name: pomerium-verify
      namespace: default
      port: 8000
  - filters:
    - type: ExtensionRef
      extensionRef:
        group: gateway.pomerium.io
        kind: PolicyFilter
        name: allow-internal
```

```json
{
  "@type": "type.googleapis.com/pomerium.config.Config",
  "routes": [
    {
      "to": ["http://pomerium-verify.default.svc.cluster.local:8000"],
      "from": "https://verify.localhost.pomerium.io",
      "prefix": "/",
      "preserveHostHeader": true,
      "loadBalancingWeights": [1]
    },
    {
      "from": "https://verify.localhost.pomerium.io",
      "prefix": "/",
      "response": { "body": "no valid backend", "status": 500 },
      "preserveHostHeader": true
    }
  ],
  "settings": {}
}
```

Reply:
I think it's correct that it maps to 2 routes, yet the response should be different:

from https://gateway-api.sigs.k8s.io/api-types/httproute/#backendrefs-optional
> BackendRefs defines API objects where matching requests should be sent. If unspecified, the rule performs no forwarding. If unspecified and no filters are specified that would result in a response being sent, a 404 error code is returned.

Reply:
Thanks for the reference. I've added a check to set a 404 response instead if there are no backendRefs at all.